commit c89308fc7a0021a9d194c796c3e29a216c82d8c6
Author: Umaro <umaro@narshe.net>
Date:   Tue Jun 21 21:57:34 2011 +0200

    fs: add hole punching to fallocate
    
        Hole punching has already been implemented by XFS and OCFS2, and has the
        potential to be implemented on both BTRFS and EXT4 so we need a generic way to
        get to this feature.  The simplest way in my mind is to add FALLOC_FL_PUNCH_HOLE
        to fallocate() since it already looks like the normal fallocate() operation.
        I've tested this patch with XFS and BTRFS to make sure XFS did what it's
        supposed to do and that BTRFS failed like it was supposed to.  Thank you,
    
        Signed-off-by: Josef Bacik <josef@redhat.com>
        Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>

diff --git a/fs/open.c b/fs/open.c
index 5463266..adefc7d 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -223,7 +223,12 @@ int do_fallocate(struct file *file, int mode, loff_t offset, loff_t len)
 		return -EINVAL;
 
 	/* Return error if mode is not supported */
-	if (mode && !(mode & FALLOC_FL_KEEP_SIZE))
+	if (mode & ~(FALLOC_FL_KEEP_SIZE | FALLOC_FL_PUNCH_HOLE))
+		return -EOPNOTSUPP;
+
+	/* Punch hole must have keep size set */
+	if ((mode & FALLOC_FL_PUNCH_HOLE) &&
+	    !(mode & FALLOC_FL_KEEP_SIZE))
 		return -EOPNOTSUPP;
 
 	if (!(file->f_mode & FMODE_WRITE))
diff --git a/include/linux/falloc.h b/include/linux/falloc.h
index 3c15510..73e0b62 100644
--- a/include/linux/falloc.h
+++ b/include/linux/falloc.h
@@ -2,6 +2,7 @@
 #define _FALLOC_H_
 
 #define FALLOC_FL_KEEP_SIZE	0x01 /* default is extend size */
+#define FALLOC_FL_PUNCH_HOLE	0x02 /* de-allocates range */
 
 #ifdef __KERNEL__
 
